[Model Averaging] Fix post_localSGD_optimizer

wayi1 opened the pull request:
I find that the original implementation of `post_localSGD_optimizer.step()` is incorrect:

Whenever `averager.average_parameters()` is called, the built-in step counter will be increased. Therefore, this should only be called exactly once per `optimizer.step()`. However, if a model has multiple param groups or params, the current implementation will call `averager.average_parameters()` multiple times and over-increase the step counter.

Relevant proposals since hierarchical SGD can be supported on `post_localSGD_optimizer`: https://github.com/pytorch/pytorch/issues/73382, https://github.com/pytorch/pytorch/issues/71325

wayi1 committed:
Fix post_localSGD_optimizer

wayi1 committed:
Update the unit test

mrshenli reviewed (approved):
LGTM

mrshenli reviewed (changes requested):
Test failure is real

```
  test_post_localSGD_optimizer_parity (__main__.TestDistBackendWithSpawn) ... INFO:numba.cuda.cudadrv.driver:init
INFO:torch.testing._internal.common_distributed:Started process 0 with pid 127986
INFO:torch.testing._internal.common_distributed:Started process 1 with pid 127987
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:1 to store for rank: 0
INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:1 to store for rank: 1
INFO:torch.distributed.distributed_c10d:Rank 0: Completed store-based barrier for key:store_based_barrier_key:1 with 2 nodes.
INFO:torch.distributed.distributed_c10d:Rank 1: Completed store-based barrier for key:store_based_barrier_key:1 with 2 nodes.
INFO:torch.testing._internal.common_distributed:Starting event listener thread for rank 1
INFO:torch.testing._internal.common_distributed:Starting event listener thread for rank 0
ERROR:torch.testing._internal.common_distributed:Caught exception: 
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 601, in run_test
    getattr(self, test_name)()
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 486, in wrapper
    fn()
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 131, in wrapper
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/distributed/distributed_test.py", line 4810, in test_post_localSGD_optimizer_parity
    averager2 = copy.deepcopy(averager)
  File "/opt/conda/lib/python3.7/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/opt/conda/lib/python3.7/copy.py", line 281, in _reconstruct
    state = deepcopy(state, memo)
  File "/opt/conda/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/opt/conda/lib/python3.7/copy.py", line 241, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/opt/conda/lib/python3.7/copy.py", line 169, in deepcopy
    rv = reductor(4)
TypeError: can't pickle torch._C._distributed_c10d.ProcessGroupNCCL objects
 exiting process 1 with exit code: 10
ERROR:torch.testing._internal.common_distributed:Caught exception: 
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 601, in run_test
    getattr(self, test_name)()
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 486, in wrapper
    fn()
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 131, in wrapper
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/distributed/distributed_test.py", line 4810, in test_post_localSGD_optimizer_parity
    averager2 = copy.deepcopy(averager)
  File "/opt/conda/lib/python3.7/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/opt/conda/lib/python3.7/copy.py", line 281, in _reconstruct
    state = deepcopy(state, memo)
  File "/opt/conda/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/opt/conda/lib/python3.7/copy.py", line 241, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/opt/conda/lib/python3.7/copy.py", line 169, in deepcopy
    rv = reductor(4)
TypeError: can't pickle torch._C._distributed_c10d.ProcessGroupNCCL objects
 exiting process 0 with exit code: 10
Process 0 terminated with exit code 10, terminating remaining processes.
ERROR (4.301s)
    test_post_localSGD_optimizer_parity errored - num_retries_left: 3
```

wayi1 committed:
fix

wayi1 commented:
> Test failure is real
> 
> ```
>   test_post_localSGD_optimizer_parity (__main__.TestDistBackendWithSpawn) ... INFO:numba.cuda.cudadrv.driver:init
> INFO:torch.testing._internal.common_distributed:Started process 0 with pid 127986
> INFO:torch.testing._internal.common_distributed:Started process 1 with pid 127987
> INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:1 to store for rank: 0
> INFO:torch.distributed.distributed_c10d:Added key: store_based_barrier_key:1 to store for rank: 1
> INFO:torch.distributed.distributed_c10d:Rank 0: Completed store-based barrier for key:store_based_barrier_key:1 with 2 nodes.
> INFO:torch.distributed.distributed_c10d:Rank 1: Completed store-based barrier for key:store_based_barrier_key:1 with 2 nodes.
> INFO:torch.testing._internal.common_distributed:Starting event listener thread for rank 1
> INFO:torch.testing._internal.common_distributed:Starting event listener thread for rank 0
> ERROR:torch.testing._internal.common_distributed:Caught exception: 
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 601, in run_test
>     getattr(self, test_name)()
>   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 486, in wrapper
>     fn()
>   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 131, in wrapper
>     return func(*args, **kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/distributed/distributed_test.py", line 4810, in test_post_localSGD_optimizer_parity
>     averager2 = copy.deepcopy(averager)
>   File "/opt/conda/lib/python3.7/copy.py", line 180, in deepcopy
>     y = _reconstruct(x, memo, *rv)
>   File "/opt/conda/lib/python3.7/copy.py", line 281, in _reconstruct
>     state = deepcopy(state, memo)
>   File "/opt/conda/lib/python3.7/copy.py", line 150, in deepcopy
>     y = copier(x, memo)
>   File "/opt/conda/lib/python3.7/copy.py", line 241, in _deepcopy_dict
>     y[deepcopy(key, memo)] = deepcopy(value, memo)
>   File "/opt/conda/lib/python3.7/copy.py", line 169, in deepcopy
>     rv = reductor(4)
> TypeError: can't pickle torch._C._distributed_c10d.ProcessGroupNCCL objects
>  exiting process 1 with exit code: 10
> ERROR:torch.testing._internal.common_distributed:Caught exception: 
> Traceback (most recent call last):
>   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 601, in run_test
>     getattr(self, test_name)()
>   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 486, in wrapper
>     fn()
>   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 131, in wrapper
>     return func(*args, **kwargs)
>   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/distributed/distributed_test.py", line 4810, in test_post_localSGD_optimizer_parity
>     averager2 = copy.deepcopy(averager)
>   File "/opt/conda/lib/python3.7/copy.py", line 180, in deepcopy
>     y = _reconstruct(x, memo, *rv)
>   File "/opt/conda/lib/python3.7/copy.py", line 281, in _reconstruct
>     state = deepcopy(state, memo)
>   File "/opt/conda/lib/python3.7/copy.py", line 150, in deepcopy
>     y = copier(x, memo)
>   File "/opt/conda/lib/python3.7/copy.py", line 241, in _deepcopy_dict
>     y[deepcopy(key, memo)] = deepcopy(value, memo)
>   File "/opt/conda/lib/python3.7/copy.py", line 169, in deepcopy
>     rv = reductor(4)
> TypeError: can't pickle torch._C._distributed_c10d.ProcessGroupNCCL objects
>  exiting process 0 with exit code: 10
> Process 0 terminated with exit code 10, terminating remaining processes.
> ERROR (4.301s)
>     test_post_localSGD_optimizer_parity errored - num_retries_left: 3
> ```

Thanks! The deepcopy of an averager now is avoided, and this should be able to fix the error.

wayi1 committed:
sync

wayi1 commented:
> Hey @wayi1, could you please rebase and address the merge conflict? Thanks!

@mrshenli should be fixed

wayi1 committed:
fix lint

wayi1 committed:
refactor

wayi1 committed:
update

mrshenli commented:
@pytorchbot land this please

mrshenli commented:
@pytorchbot merge this please
